stl/inc/coroutine

```diff
@@ -0,0 +1,168 @@
+// coroutine standard header (core)
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef _COROUTINE_
+#define _COROUTINE_
+#include <yvals_core.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#ifdef _RESUMABLE_FUNCTIONS_SUPPORTED
+#pragma message("The contents of <coroutine> are not available with /await. Use <experimental/coroutine> instead.")
+#else // ^^^ /await ^^^ / vvv no /await vvv
+#ifndef __cpp_lib_coroutine
+#pragma message("The contents of <coroutine> are available only with C++20 or later.")
+#else // ^^^ __cpp_lib_coroutine not defined / __cpp_lib_coroutine defined vvv
+#ifndef _ALLOW_COROUTINE_ABI_MISMATCH
+#pragma detect_mismatch("_COROUTINE_ABI", "2")
+#endif // _ALLOW_COROUTINE_ABI_MISMATCH
+
+#include <compare>
+#include <type_traits>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#undef new
+
+_STD_BEGIN
+
+// STRUCT TEMPLATE coroutine_traits
+template <class _Ret, class = void>
+struct _Coroutine_traits {};
+
+template <class _Ret>
+struct _Coroutine_traits<_Ret, void_t<typename _Ret::promise_type>> {
+    using promise_type = typename _Ret::promise_type;
+};
+
+template <class _Ret, class...>
+struct coroutine_traits : _Coroutine_traits<_Ret> {};

```

Comment:
Just thinking out loud. 

Would it improve the error messages when the detection of the `promise_type` would be done via a concept?

Reply:
I don't think there would be any observable difference. The partial specialization either matches or it doesn't, there's never a "you didn't match this partial specialization because" diagnostic.

Using concepts (or more precisely requires-clauses) _would_ let us avoid the `_Coroutine_traits` helper, we can write:
```c++
template <class, class...>
struct coroutine_traits {};

template <class _Ret>
    requires { typename _Ret::promise_type; }
struct coroutine_traits<_Ret> {
    using promise_type = typename _Ret::promise_type;
};
```
but it's not worth breaking IntelliSense (which doesn't do Concepts yet) to make the code a bit shorter.

Reply:
Yeah, breaking IntelliSense would be bad. 

Are you missing a second `requires` in your example?

Reply:
> Are you missing a second `requires` in your example?

Yes; I'm missing the parameter pack, too. Let's try again:
```c++
template <class, class...>
struct coroutine_traits {};

template <class _Ret, class... _Ts>
    requires requires { typename _Ret::promise_type; }
struct coroutine_traits<_Ret, _Ts...> {
    using promise_type = typename _Ret::promise_type;
};
```